File: analytical_engine/java/giraph-on-grape/src/main/java/org/apache/giraph/time/Time.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.giraph.time;

import java.util.Date;

/**
 * Interface for handling Time related operations so that they can be mocked
 * for testing.
 */
public interface Time {
    /** Microseconds per millisecond */
    long US_PER_MS = 1000;
    /** Nanoseconds per microsecond */
    long NS_PER_US = 1000;
    /** Nanoseconds per millisecond */
    long NS_PER_MS = US_PER_MS * NS_PER_US;
    /** Milliseconds per second */
    long MS_PER_SECOND = 1000;
    /** Milliseconds per second (as float) */
    float MS_PER_SECOND_AS_FLOAT = MS_PER_SECOND * 1f;
    /** Microseconds per second */
    long US_PER_SECOND = US_PER_MS * MS_PER_SECOND;
    /** Microseconds per second (as float) */
    float US_PER_SECOND_AS_FLOAT = US_PER_SECOND * 1f;
    /** Nanoseconds per second */
    long NS_PER_SECOND = NS_PER_US * US_PER_SECOND;
    /** Nanoseconds per second (as float) */
    float NS_PER_SECOND_AS_FLOAT = NS_PER_SECOND * 1f;
    /** Seconds per hour */
    long SECONDS_PER_HOUR = 60 * 60;
    /** Seconds per day */
    long SECONDS_PER_DAY = 24 * SECONDS_PER_HOUR;
    /** Milliseconds per hour */
    long MS_PER_HOUR = SECONDS_PER_HOUR * MS_PER_SECOND;
    /** Milliseconds per day */
    long MS_PER_DAY = SECONDS_PER_DAY * MS_PER_SECOND;

    /**
     * Get the current milliseconds
     *
     * @return The difference, measured in milliseconds, between
     *         the current time and midnight, January 1, 1970 UTC.
     */
    long getMilliseconds();

    /**
     * Get the current microseconds
     *
     * @return The difference, measured in microseconds, between
     *         the current time and midnight, January 1, 1970 UTC.
     */
    long getMicroseconds();

    /**
     * Get the current nanoseconds
     *
     * @return The difference, measured in nanoseconds, between
     *         the current time and midnight, January 1, 1970 UTC.
     */
    long getNanoseconds();

    /**
     * Get the current seconds
     *
     * @return The difference, measured in seconds, between
     *         the current time and midnight, January 1, 1970 UTC.
     */
    int getSeconds();

    /**
     * Get the current date
     *
     * @return Current date
     */
    Date getCurrentDate();

    /**
     * Current thread should sleep for some number of milliseconds.
     *
     * @param milliseconds Milliseconds to sleep for
     * @throws InterruptedException
     */
    void sleep(long milliseconds) throws InterruptedException;
}

File: interactive_engine/rust-common/src/util/time.rs
//
//! Copyright 2020 Alibaba Group Holding Limited.
//! 
//! Licensed under the Apache License, Version 2.0 (the "License");
//! you may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//! 
//!     http://www.apache.org/licenses/LICENSE-2.0
//! 
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

use std::time::{SystemTime, UNIX_EPOCH, Duration, Instant};
use std::thread;
use std::sync::atomic::{AtomicUsize, Ordering};

pub fn current_time_millis() -> u64 {
    let now = SystemTime::now();
    let d = now.duration_since(UNIX_EPOCH).unwrap();
    d.as_secs() * 1000 + (d.subsec_nanos() / 1000000) as u64
}

pub fn current_time_secs() -> u64 {
    let now = SystemTime::now();
    now.duration_since(UNIX_EPOCH).unwrap().as_secs()
}

pub fn duration_to_millis(d: &Duration) -> f64 {
    d.as_secs() as f64 * 1000.0 + (d.subsec_nanos() as f64 / 1000000.0)
}

pub fn duration_to_nanos(d: &Duration) -> u64 {
    d.as_secs()  * 1000_000_000 + d.subsec_nanos() as u64
}

pub fn sleep_ms(ms: u64) {
    thread::sleep(Duration::from_millis(ms));
}

pub fn sleep_until(flag: &AtomicUsize, target: usize) -> SleepGuard {
    while flag.load(Ordering::Relaxed) != target {
        sleep_ms(100);
    }
    SleepGuard::new(flag, target)
}

pub struct SleepGuard<'a> {
    flag: &'a AtomicUsize,
    target: usize,
}

impl<'a> SleepGuard<'a> {
    fn new(flag: &'a AtomicUsize, target: usize) -> Self {
        SleepGuard {
            flag,
            target,
        }
    }
}

impl<'a> Drop for SleepGuard<'a> {
    fn drop(&mut self) {
        self.flag.store(self.target + 1, Ordering::Relaxed);
    }
}

pub struct Timer {
    timer: Instant,
}

impl Timer {
    pub fn new() -> Self {
        Timer {
            timer: Instant::now(),
        }
    }

    pub fn elapsed_ms(&self) -> f64 {
        let t = self.timer.elapsed();
        t.as_secs() as f64 * 1000.0 + t.subsec_nanos() as f64 / 1000000.0
    }

    pub fn elasped_secs(&self) -> f64 {
        let t = self.timer.elapsed();
        t.as_secs() as f64 + t.subsec_nanos() as f64 / 1000000000.0
    }
}
